compiler/q-implant/src/QImplant.cpp

```diff
@@ -282,4 +285,38 @@ void QImplant::write(loco::Graph *g)
   }
 }
 
+void QImplant::forward_qparam(loco::Graph *g)
+{
+  std::set<luci::CircleOpcode> forwardable_opcode{
+    luci::CircleOpcode::RESHAPE, luci::CircleOpcode::SPLIT, luci::CircleOpcode::TRANSPOSE};
```

Comment:
plz add comment why only three Ops are listed here

Reply:
(optional) use `forwardable_opcode.emplace()` to add these three for future extendability

Reply:
During model conversion(onnx->tf) we faced only these operators until now.
Honestly not sure about which operators can be added on model conversion period and can use predecessor quantization parameter.

When found/faced others, we will add here.

At first I put elements vertically, and ran `nnas format'. I surprised that `nnas` make it horizontally.
So using `emplace` would be best option, because we can read vertically and can extend list easily.

Lastly, `add comment` means comment in here or add comment on `cpp`?

Reply:
> Honestly not sure about which operators can be added on model conversion period

Then, plz add decision factors how you selected these Ops.

> When found/faced others, we will add here.

You mean by until SSAFY?

> Lastly, add comment means comment in here or add comment on cpp?

In the cpp source code.

Reply:
>Then, plz add decision factors how you selected these Ops.

- Operators can be added during model conversion
- Operators are not quantiazable, so can pass input tensor's quantization parameter to output tensor.

It's our thoughts, would you think It can be the decision factor what you mentioned?

> You mean by until SSAFY?

Not sure about, but maybe..?
If I have time to look `onnx-tf`, I will find if any other operators can be added.

> In the cpp source code.

If the decision factor above is appropriate, I will place it to cpp with your other suggestions and commit it.

Thank you.

Reply:
> It's our thoughts, would you think It can be the decision factor what you mentioned?

nope. how did you added `RESHAPE` ?

Reply:
I'm not sure how to answer because knowledge is limited.

- Operators can be added during model conversion
As we remember, we saw reshape was added on `onnx` to `tf` conversion, but if you suspicious of it, we can't answer about that.
(Don't have much knowledges, saw it in low version of torch/onnx-tf so not know it was wrong conversion or not, can't remember which model conversion make it)

[Reshape](https://www.tensorflow.org/api_docs/python/tf/reshape)
Reshape have input tensor, and shape tensor(in int), so not quantizable.
As we know, reshape operator don't change quantization paramter. (input/output tensor's quantization parameter same)

Last, if `qparam.json` contain quantization parameter(original model in other frame have reshape operator) of reshape operator, it would not work (because it have quantization parameter).

Reply:
in https://github.com/Samsung/ONE/pull/11593#discussion_r1336867644

> If I understand PropagateQParamForwardPass correctly, running PropagateQParamForwardPass will propagate quantization parameter when operations like Reshape, Transpose.

what was your understanding?

Reply:
When reading the code, `PropagateQParamForwardPass` will copy predecessor node's quantization parameter to node like reshape, transpose.

So I understood that if I run the `PropagateQParamForwardPass`, some operators like Reshape or Transpose can get quantization parameter from predecessor.

> Then, plz add decision factors how you selected these Ops.

Is asking why there are three operators in list

> how did you added RESHAPE ?

Is saying that we should exclude reshape because reshape operator would not be added at model conversion(onnx -> tf) period. (What we saw/remember could be wrong, and I'm not 100% sure whether reshape can be added or not on model conversion)

Reply:
> When reading the code, PropagateQParamForwardPass will copy ...

Do you mean, the code wrote by @jinevening ?

Reply:
Yes, when I reading `PropagateQParamForwardPass` on luci/pass, it looks like copying quantization parameter when some operator like reshape, transpose.
So `PropagateQParamForwardPass` was thought to be used in `q-implant`, but when looked closely,

> copy_qparam assume that source and destination node have same scale,zerop size, so we need to assign source's size into destination, right?

`copy_qparam` on `PropagateQParamForwardPass` assume that destination model already have same scale/zero point size. `q-implant` could not have quantization parameter (assume using it to not quantized model to make quantized model), so we have to assign proper size for not quantized model.

> when I run the code by using PropagateQParamForwardPass, exception occured due to remaining float. (Quantization Parameter copied, but dtype not changed)

When used it, dtype not copied, and need to find another solution.

> It seems that you used queue just to traverse the graph. We already have helper functions for that purpose, ex: loco::postorder_traversal or Pass. Could you use them?

@jinevening recommended to use postorder_traversal.

> what was your understanding?

That's what I understood, and not sure about relation between `PropagateQParamForwardPass` and `Reshape`

Reply:
> ... relation between PropagateQParamForwardPass and Reshape

RESHAPE was the first item in the code, so I asked the reason, it was not about specific `Reshape`.
You should have answered like "I don't know cause the original code was written from @jinevening".

Anyway, my suggestion is like
```cxx
  // TODO: add comment about how to add to the set
  std::set<luci::CircleOpcode> forwardable_opcode;
  forwardable_opcode.emplace(luci::CircleOpcode::RESHAPE);
  forwardable_opcode.emplace(luci::CircleOpcode::SPLIT);
  forwardable_opcode.emplace(luci::CircleOpcode::TRANSPOSE);
  // TODO add more Ops
```

Reply:
>nope. how did you added RESHAPE ?

I understood it as why did we added `reshape` itself.

Firstly, we didn't know about `luci::PropagateQParamForwardPass` until @jinevening recognized what we want to do is same/similar with  `PropagateQParamForwardPass` and asked to check possible to use to prevent duplicated code yesterday.

> Honestly not sure about which operators can be added on model conversion period

What I'm not sure about, is don't know all operators can be added during model conversion, so we placed what we saw on conversion for now. 

>>Then, plz add decision factors how you selected these Ops.

>  Operators can be added during model conversion
 Operators are not quantiazable, so can pass input tensor's quantization parameter to output tensor.

These are 'as is' how we selected, `decision factors` may mean which operators can be placed on list (common attribute that operators in list must satisfied).
- Operator does not change quantization parameter. (min, max, zero point, scale)
- reshape, transpose and split are just change tensor value's position/dimension, and not change individual value so quantization parameter does not change.
- So input tensor's quantization parameter and output tensor's quantization parameter must be same.


That's what all I know, not sure that answer ,is what you hoped, or not.


>Anyway, my suggestion is like

And I think it is good for people to read and expand, I will reflect it.

Thank you.

Reply:
> and not change individual value so quantization parameter does not change.


@jinevening , does this hold for channel wise?

Reply:
> does this hold for channel wise?

Yes. Activation is quantized layer-wise, not channel-wise.